py-rse/provenance.Rmd

```diff
@@ -0,0 +1,381 @@
+    (i.e. a step-by-step account of how the software and scripts were actually used).
+
+Unfortunately,
+librarians, publishers, and regulatory bodies are still trying to determine
+the best way to document and archive material like this,
+so there is not yet anything like the FAIR Principles.
+The best advice we can give is presented below.
+
+### Software Description
+
+FIXME: This section would make more sense if we were using a conda environment
+from the beginning (i.e. in the installation instructions).
+We should consider the merits of doing that.
+
+In order to document the software packages that were used in our analysis,
+the bare minimum requirement is to archive a list of
+the names and version numbers of each software package.
+We can get version information for the Python packages we are using by running:
+
+```shell
+$ pip freeze
+```
+
+```text
+FIXME: Show output
+```
+
+Other command line tools will often have an option like `--version` or `--status`
+to access the version information.
+
+Archiving a list of package names and version numbers would mean that our
+software environment is technically reproducible,
+but it would be left up to the reader of the report
+to figure out how to get all those packages installed and working together.
+This might be fine for a small number of packages with very few dependencies,
+but in more complex cases we probably want to make life easier for the reader
+(and for our future selves looking to re-run the analysis).
+One way to make things easier is to export a description of a complete
+`conda` environment, which can be saved as YAML using:
+
+```shell
+$ conda env export > environment.yml
+```
+
+That software environment can then be recreated on another computer
+with just one line of code:
+
+```
+$ conda env create -f environment.yml
+```
+
+More complex tools like [Docker][docker]
+can literally install our entire environment
+(down to the precise version of the operating system)
+on a different computer @Nust2020.
+However,
+their complexity can be daunting,
+and there is a lot of debate
+about how well (or whether) they actually make research more reproducible in practice.
+
+### Analysis Scripts
+
+Later in this book we will package and release our Zipf's Law code
+so that it can be downloaded and installed by the wider research community
+just like any other Python package (Chapter \@ref(packaging)).
+Going through the process of formally packaging your code can be a good option
+if there's a user community out there interested in using and/or building it,
+but often the scripts and notebooks we write
+to produce a particular figure or table
+will not be of broad interest.
+In other words,
+these scripts were written to get a specific job done
+(e.g. to produce Figure 2b in a report)
+and do not have a wider application.
+To fully capture the provenance of the results presented in a report,
+these analysis scripts and/or notebooks can simply be collected up into a zip folder
```

Comment:
I'd suggest instead of zipping, suggest that analysis scripts/notebooks be given a Git tag, not put into a zip folder. Git tags are much easier to access and can also be created into zip folders if needed. Plus the Git tag links with the version control content.